File: projects/interpreter/lexer.go
package main

// Lexer performs lexical analysis on source code, converting it into tokens.
type Lexer struct {
	input   string
	pos     int    // current position in input (points to current char)
	readPos int    // current reading position (after current char)
	ch      byte   // current char under examination
	line    int    // current line number
	col     int    // current column number
}

// NewLexer creates a new Lexer for the given input string.
func NewLexer(input string) *Lexer {
	l := &Lexer{input: input, line: 1, col: 0}
	l.readChar()
	return l
}

// readChar reads the next character and advances the position.
func (l *Lexer) readChar() {
	if l.readPos >= len(l.input) {
		l.ch = 0 // ASCII NUL = EOF
	} else {
		l.ch = l.input[l.readPos]
	}
	l.pos = l.readPos
	l.readPos++
	if l.ch == '\n' {
		l.line++
		l.col = 0
	} else {
		l.col++
	}
}

// peekChar returns the next character without advancing.
func (l *Lexer) peekChar() byte {
	if l.readPos >= len(l.input) {
		return 0
	}
	return l.input[l.readPos]
}

// NextToken scans the input and returns the next token.
func (l *Lexer) NextToken() Token {
	var tok Token

	l.skipWhitespace()

	// Track token start position
	line := l.line
	col := l.col

	switch l.ch {
	case '+':
		tok = Token{Type: PLUS, Literal: string(l.ch), Line: line, Col: col}
	case '-':
		tok = Token{Type: MINUS, Literal: string(l.ch), Line: line, Col: col}
	case '*':
		tok = Token{Type: STAR, Literal: string(l.ch), Line: line, Col: col}
	case '/':
		if l.peekChar() == '/' {
			// Single-line comment
			l.skipComment()
			return l.NextToken()
		}
		tok = Token{Type: SLASH, Literal: string(l.ch), Line: line, Col: col}
	case '%':
		tok = Token{Type: PERCENT, Literal: string(l.ch), Line: line, Col: col}
	case '(':
		tok = Token{Type: LPAREN, Literal: string(l.ch), Line: line, Col: col}
	case ')':
		tok = Token{Type: RPAREN, Literal: string(l.ch), Line: line, Col: col}
	case '{':
		tok = Token{Type: LBRACE, Literal: string(l.ch), Line: line, Col: col}
	case '}':
		tok = Token{Type: RBRACE, Literal: string(l.ch), Line: line, Col: col}
	case ',':
		tok = Token{Type: COMMA, Literal: string(l.ch), Line: line, Col: col}
	case ';':
		tok = Token{Type: SEMICOL, Literal: string(l.ch), Line: line, Col: col}
	case '=':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: EQ, Literal: string(ch) + string(l.ch), Line: line, Col: col}
		} else {
			tok = Token{Type: ASSIGN, Literal: string(l.ch), Line: line, Col: col}
		}
	case '!':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: NEQ, Literal: string(ch) + string(l.ch), Line: line, Col: col}
		} else {
			tok = Token{Type: ILLEGAL, Literal: string(l.ch), Line: line, Col: col}
		}
	case '<':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: LTE, Literal: string(ch) + string(l.ch), Line: line, Col: col}
		} else {
			tok = Token{Type: LT, Literal: string(l.ch), Line: line, Col: col}
		}
	case '>':
		if l.peekChar() == '=' {
			ch := l.ch
			l.readChar()
			tok = Token{Type: GTE, Literal: string(ch) + string(l.ch), Line: line, Col: col}
		} else {
			tok = Token{Type: GT, Literal: string(l.ch), Line: line, Col: col}
		}
	case '"':
		tok.Type = STRING
		tok.Literal = l.readString()
		tok.Line = line
		tok.Col = col
	case 0:
		tok = Token{Type: EOF, Literal: "", Line: line, Col: col}
	default:
		if isLetter(l.ch) {
			tok.Literal = l.readIdentifier()
			tok.Type = LookupIdent(tok.Literal)
			tok.Line = line
			tok.Col = col
			return tok
		} else if isDigit(l.ch) {
			tok.Type = NUMBER
			tok.Literal = l.readNumber()
			tok.Line = line
			tok.Col = col
			return tok
		} else {
			tok = Token{Type: ILLEGAL, Literal: string(l.ch), Line: line, Col: col}
		}
	}

	l.readChar()
	return tok
}

// readIdentifier reads an identifier and advances the position.
func (l *Lexer) readIdentifier() string {
	pos := l.pos
	for isLetter(l.ch) || isDigit(l.ch) {
		l.readChar()
	}
	return l.input[pos:l.pos]
}

// readNumber reads a number (integer or float) and advances the position.
func (l *Lexer) readNumber() string {
	pos := l.pos
	for isDigit(l.ch) {
		l.readChar()
	}
	if l.ch == '.' && isDigit(l.peekChar()) {
		l.readChar() // consume '.'
		for isDigit(l.ch) {
			l.readChar()
		}
	}
	return l.input[pos:l.pos]
}

// readString reads a string literal and advances the position.
func (l *Lexer) readString() string {
	l.readChar() // consume opening "
	pos := l.pos
	for l.ch != '"' && l.ch != 0 {
		if l.ch == '\\' {
			l.readChar() // skip escape char
		}
		l.readChar()
	}
	str := l.input[pos:l.pos]
	// Process escape sequences
	result := ""
	for i := 0; i < len(str); i++ {
		if str[i] == '\\' && i+1 < len(str) {
			switch str[i+1] {
			case 'n':
				result += "\n"
			case 't':
				result += "\t"
			case '\\':
				result += "\\"
			case '"':
				result += "\""
			default:
				result += string(str[i]) + string(str[i+1])
			}
			i++ // skip next char
		} else {
			result += string(str[i])
		}
	}
	return result
}

// skipWhitespace skips whitespace characters.
func (l *Lexer) skipWhitespace() {
	for l.ch == ' ' || l.ch == '\t' || l.ch == '\n' || l.ch == '\r' {
		l.readChar()
	}
}

// skipComment skips a single-line comment (// to end of line).
func (l *Lexer) skipComment() {
	for l.ch != '\n' && l.ch != 0 {
		l.readChar()
	}
}

// isLetter checks if a byte is a letter, underscore, or non-ASCII character.
func isLetter(ch byte) bool {
	return 'a' <= ch && ch <= 'z' || 'A' <= ch && ch <= 'Z' || ch == '_'
}

// isDigit checks if a byte is a digit.
func isDigit(ch byte) bool {
	return '0' <= ch && ch <= '9'
}

// Tokenize returns all tokens from the input.
func Tokenize(input string) []Token {
	lexer := NewLexer(input)
	var tokens []Token
	for {
		tok := lexer.NextToken()
		tokens = append(tokens, tok)
		if tok.Type == EOF {
			break
		}
	}
	return tokens
}

File: projects/interpreter/python/src/lexer.py
"""
词法分析器（Lexer）

将源代码字符流转换为Token流。
使用有限状态机逐字符扫描，识别关键字、标识符、数字、字符串等。
"""

from .token import Token, TokenType


class LexerError(Exception):
    """词法分析错误"""

    def __init__(self, message: str, line: int, column: int):
        super().__init__(f"[行 {line}, 列 {column}] 词法错误: {message}")
        self.line = line
        self.column = column


class Lexer:
    """
    词法分析器

    将源代码字符串转换为Token序列。
    支持：
    - 数字（整数和浮点数）
    - 字符串（双引号）
    - 标识符和关键字
    - 运算符和分隔符
    - 单行注释（//）和多行注释（/* */）
    """

    def __init__(self, source: str):
        self.source = source
        self.pos = 0           # 当前字符位置
        self.line = 1          # 当前行号
        self.column = 1        # 当前列号
        self.tokens: list[Token] = []

    def tokenize(self) -> list[Token]:
        """将源代码转换为Token列表"""
        while not self._is_at_end():
            self._skip_whitespace_and_comments()
            if self._is_at_end():
                break
            self._read_token()

        # 添加EOF token
        self.tokens.append(Token(TokenType.EOF, "", self.line, self.column))
        return self.tokens

    def _is_at_end(self) -> bool:
        return self.pos >= len(self.source)

    def _peek(self) -> str:
        """查看当前字符（不消耗）"""
        if self._is_at_end():
            return "\0"
        return self.source[self.pos]

    def _peek_next(self) -> str:
        """查看下一个字符"""
        if self.pos + 1 >= len(self.source):
            return "\0"
        return self.source[self.pos + 1]

    def _advance(self) -> str:
        """消耗并返回当前字符"""
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.column = 1
        else:
            self.column += 1
        return ch

    def _expect(self, expected: str) -> bool:
        """如果下一个字符匹配则消耗它"""
        if self._is_at_end() or self.source[self.pos] != expected:
            return False
        self._advance()
        return True

    def _add_token(self, token_type: TokenType, literal: str, line: int, column: int):
        """添加Token到列表"""
        self.tokens.append(Token(token_type, literal, line, column))

    def _skip_whitespace_and_comments(self):
        """跳过空白字符和注释"""
        while not self._is_at_end():
            ch = self._peek()
            if ch in (" ", "\t", "\r", "\n"):
                self._advance()
            elif ch == "/" and self._peek_next() == "/":
                # 单行注释
                while not self._is_at_end() and self._peek() != "\n":
                    self._advance()
            elif ch == "/" and self._peek_next() == "*":
                # 多行注释
                self._advance()  # /
                self._advance()  # *
                while not self._is_at_end():
                    if self._peek() == "*" and self._peek_next() == "/":
                        self._advance()  # *
                        self._advance()  # /
                        break
                    self._advance()
            else:
                break

    def _read_token(self):
        """读取下一个Token"""
        line, column = self.line, self.column
        ch = self._advance()

        match ch:
            # 单字符Token
            case "+":
                if self._expect("="):
                    self._add_token(TokenType.PLUS_ASSIGN, "+=", line, column)
                else:
                    self._add_token(TokenType.PLUS, "+", line, column)
            case "-":
                if self._expect("="):
                    self._add_token(TokenType.MINUS_ASSIGN, "-=", line, column)
                else:
                    self._add_token(TokenType.MINUS, "-", line, column)
            case "*":
                if self._expect("*"):
                    self._add_token(TokenType.POWER, "**", line, column)
                else:
                    self._add_token(TokenType.STAR, "*", line, column)
            case "/":
                self._add_token(TokenType.SLASH, "/", line, column)
            case "%":
                self._add_token(TokenType.PERCENT, "%", line, column)
            case "(":
                self._add_token(TokenType.LPAREN, "(", line, column)
            case ")":
                self._add_token(TokenType.RPAREN, ")", line, column)
            case "{":
                self._add_token(TokenType.LBRACE, "{", line, column)
            case "}":
                self._add_token(TokenType.RBRACE, "}", line, column)
            case "[":
                self._add_token(TokenType.LBRACKET, "[", line, column)
            case "]":
                self._add_token(TokenType.RBRACKET, "]", line, column)
            case ",":
                self._add_token(TokenType.COMMA, ",", line, column)
            case ";":
                self._add_token(TokenType.SEMICOLON, ";", line, column)
            case ":":
                self._add_token(TokenType.COLON, ":", line, column)
            case ".":
                self._add_token(TokenType.DOT, ".", line, column)

            # 可能是双字符的Token
            case "=":
                if self._expect("="):
                    self._add_token(TokenType.EQ, "==", line, column)
                else:
                    self._add_token(TokenType.ASSIGN, "=", line, column)
            case "!":
                if self._expect("="):
                    self._add_token(TokenType.NOT_EQ, "!=", line, column)
                else:
                    self._add_token(TokenType.NOT, "!", line, column)
            case "<":
                if self._expect("="):
                    self._add_token(TokenType.LTE, "<=", line, column)
                else:
                    self._add_token(TokenType.LT, "<", line, column)
            case ">":
                if self._expect("="):
                    self._add_token(TokenType.GTE, ">=", line, column)
                else:
                    self._add_token(TokenType.GT, ">", line, column)
            case "&":
                if self._expect("&"):
                    self._add_token(TokenType.AND, "&&", line, column)
                else:
                    raise LexerError(f"意外字符 '&'", line, column)
            case "|":
                if self._expect("|"):
                    self._add_token(TokenType.OR, "||", line, column)
                else:
                    raise LexerError(f"意外字符 '|'", line, column)

            # 字符串
            case '"':
                self._read_string(line, column)

            # 数字
            case _ if ch.isdigit():
                self._read_number(ch, line, column)

            # 标识符/关键字
            case _ if ch.isalpha() or ch == "_":
                self._read_identifier(ch, line, column)

            case _:
                raise LexerError(f"意外字符 '{ch}'", line, column)

    def _read_string(self, line: int, column: int):
        """读取字符串字面量"""
        parts: list[str] = []
        while not self._is_at_end() and self._peek() != '"':
            if self._peek() == "\\":
                self._advance()  # 跳过反斜杠
                escape = self._advance()
                escape_map = {
                    "n": "\n", "t": "\t", "r": "\r",
                    "\\": "\\", '"': '"',
                }
                parts.append(escape_map.get(escape, f"\\{escape}"))
            elif self._peek() == "\n":
                raise LexerError("未终止的字符串", line, column)
            else:
                parts.append(self._advance())

        if self._is_at_end():
            raise LexerError("未终止的字符串", line, column)

        self._advance()  # 消耗结束的 "
        self._add_token(TokenType.STRING, "".join(parts), line, column)

    def _read_number(self, first_char: str, line: int, column: int):
        """读取数字字面量"""
        num_chars = [first_char]
        is_float = False

        while not self._is_at_end() and (self._peek().isdigit() or self._peek() == "."):
            if self._peek() == ".":
                if is_float:
                    break  # 第二个点号停止
                is_float = True
            num_chars.append(self._advance())

        self._add_token(TokenType.NUMBER, "".join(num_chars), line, column)

    def _read_identifier(self, first_char: str, line: int, column: int):
        """读取标识符或关键字"""
        chars = [first_char]
        while not self._is_at_end() and (self._peek().isalnum() or self._peek() == "_"):
            chars.append(self._advance())

        literal = "".join(chars)
        token_type = Token.lookup_ident(literal)
        self._add_token(token_type, literal, line, column)
